### PR TITLE
improved speed  of fast corner detect

### PR DESCRIPTION
**Pull request by spdfghi**
## Description
The correct implementation of Fast corner detection involves moving the fast check when n_fast >= 12 to the beginning of calculating the pixel values around the entire circle, which can significantly reduce the time spent on non-corner detection.

### CODE REVIEW

**Comment by spdfghi** [comment]
I am new to github, please help me with the label

**Comment by lagru** [comment]
Hey @spdfghi, you haven't done anything wrong. You don't have the permissions to add a label to this PR, that's usually done by our maintainers. Sorry, for the confusion on your part. :)

**Comment by lagru** [comment]
I don't really notice a speed difference when testing this locally. Could you maybe explain how this improves performance and / or provide a snippet for which your change performs faster?

It also don't understand the purpose of that entire snippet https://github.com/scikit-image/scikit-image/blob/b8ef57a285b9c09c850f56113cea45b3d734f473/skimage/feature/corner_cy.pyx#L160-L170

It looks like it has no effect and none of the modified variables `speed_sum_b` or `speed_sum_d` are ever used. Am I missing something?

**Comment by lagru** [comment]
Uff, I entirely missed that the `continue` was affecting the outer loop. Testing with 

```python
from skimage.feature import corner_fast, corner_peaks
square = np.zeros((12, 12))
square[3:9, 3:9] = 1
square.astype(int)
%timeit corner_fast(square)
```
shows a small effect indeed:

- `main` : 5.45 µs
- This PR : 5.13 µs
- Removing the snippet : 8.62 µs

Question is, does this affect the result? I currently don't understand this enough to review that..

**Comment by spdfghi** [comment]
Thank you for your feedback. The purpose of this code snippet is as follows: In the Fast algorithm, when there are consecutive n_fast points whose brightness is greater than or less than the brightness of the center point, it is considered a corner. When n_fast &gt;= 12, every 4 points are tested. If both speed_sum_b and speed_sum_d are less than 3, then there definitely aren't more than 12 consecutive points with brightness greater than or less than the center point's brightness. By placing this section at the beginning, when the image keypoints are sparse, it can skip the calculation of circle_intensities (calculating 16 points) and only calculate 4 points.
If you are using images with either a higher accuracy corner detection with n_fast=9 or very dense feature points, then there won't be any improvement in performance. If necessary, I will provide the testing code. I've been a bit busy lately, sorry...





spdfghi
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "scikit-image/scikit-image"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年4月9日(星期二) 晚上8:52
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [scikit-image/scikit-image] improved speed  of fast corner detect (PR #7379)





 
I don't really notice a speed difference when testing this locally. Could you maybe explain how this improves performance and / or provide a snippet for which your change performs faster?
 
It also don't understand the purpose of that entire snippet https://github.com/scikit-image/scikit-image/blob/b8ef57a285b9c09c850f56113cea45b3d734f473/skimage/feature/corner_cy.pyx#L160-L170
 
It looks like it has no effect and none of the modified variables speed_sum_b or speed_sum_d are ever used. Am I missing something?

File: skimage/feature/corner_cy.pyx [hunk 1]
```diff
@@ -145,6 +145,19 @@ def _corner_fast(np_floats[:, ::1] image, signed char n, np_floats threshold):
                 lower_threshold = curr_pixel - threshold
                 upper_threshold = curr_pixel + threshold
 
+                # High speed test for n >= 12
+                if n >= 12:
+                    speed_sum_b = 0
+                    speed_sum_d = 0
+                    for k in range(0, 16, 4):
+                        ring_pixel = image[i + rp[k], j + cp[k]]
```
Comment (lagru) [inline skimage/feature/corner_cy.pyx:153]: Thank you for the prompt response @spdfghi. It's beginning to make sense. A follow-up question:

Considering the offset arrays `rp` and `cp` have 16 elements, how do tehese lines make sense. Shouldn't `rp[16]` be out of bounds and return a nonsensical result? This is also an issue with the original implementation...

Comment (spdfghi) [reply skimage/feature/corner_cy.pyx:153]: you know it. Just to clarify.  I usually work with c++,  i don't know how pyx become lib. I will check letter, but i think testing code is just like this, with larger image:
`from skimage.feature import corner_fast, corner_peaks
square = np.zeros(1200, 1200))
square.astype(int)`

**Comment by lagru** [comment]
Hey @spdfghi, why did you close this? Are you unable to continue working on this PR together with us?

File: skimage/feature/corner_cy.pyx (hunk 1 above)
Comment (spdfghi) [reply skimage/feature/corner_cy.pyx:153]: @lagru i am not native English speaker, so i might misunderstood the question, as  range for in python is used like [a, b),  range(0, 16, 4) is 0, 4, 8, 12

**Pull request closed without merging**